### ISSUE TITLE
DynamicParameters built inside a PowerShell function ignore ParameterAttribute.Position property

### ISSUE THREAD

**Issue by lfshr** [PowerShell/PowerShell#3704]
DynamicParameters built inside a PowerShell function ignore ParameterAttribute.Position property

Steps to reproduce
------------------
```
function DynamParamFunc {
    [CmdletBinding()]
    param(
        [Parameter(Position = 2)]
        $Param2
    )

    dynamicparam {
        $Param1Attribute = New-Object -TypeName System.Management.Automation.ParameterAttribute
        $Param1Attribute.Position = 1

        $AttributeCollection = New-Object -TypeName System.Collections.ObjectModel.Collection[System.Attribute]
        $AttributeCollection.Add($Param1Attribute)
        $Param1Param = New-Object System.Management.Automation.RuntimeDefinedParameter('Param1', [string], $AttributeCollection)
        $ParamDictionary = New-Object System.Management.Automation.RuntimeDefinedParameterDictionary
        $ParamDictionary.Add('Param1', $Param1Param)
        return $ParamDictionary
    }

    end {
        $PSCmdlet.GetDynamicParameters()
        $Param2
    }
}
```

Try and use tab completion for the parameters, Param2 will appear first despite Param1 having a lower position.

Expected behavior
-----------------
Param1 is position 1
Param2 is position 2

Actual behavior
---------------
Param2 is position 1
Param1 is position 2

Environment data
----------------
```
Name                           Value                                                                                                                                                                                           
----                           -----                                                                                                                                                                                           
PSVersion                      5.1.14393.1066                                                                                                                                                                                  
PSEdition                      Desktop                                                                                                                                                                                         
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0...}                                                                                                                                                                         
BuildVersion                   10.0.14393.1066                                                                                                                                                                                 
CLRVersion                     4.0.30319.42000                                                                                                                                                                                 
WSManStackVersion              3.0                                                                                                                                                                                             
PSRemotingProtocolVersion      2.3                                                                                                                                                                                             
SerializationVersion           1.1.0.1  
```

**Comment by iSazonov** [comment]
@lfshr Thanks for your report! Here is PowerShell Core repo - please check the Issue with latest PowerShell Core version.

**Comment by lzybkr** [comment]
There is no intentional correlation in ordering of parameters during tab completion and the position attribute.

Generally speaking, the order is the order declared, but this is not guaranteed. For binary cmdlets, we rely on reflection, and the order returned by reflection is not guaranteed. It is often the order the parameters were defined in C#, but not necessarily.

Also note that hosts are free to reorder the completions, e.g. Intellisense may sort them alphabetically.

It's also not clear if honoring the position provides a good experience with all cmdlets - e.g. if we started mixing parameter sets to honor position, things could get more confusing.

**Comment by lfshr** [comment]
The 'Position' parameter seems fairly pointless if it doesn't position it in the way specified. Generally speaking I like to position things in the order of importance, with non-mandatory parameters being at the end. This is currently not achievable with a dynamic parameter. What if I wanted a really important parameter to help the user by offering a dynamic ValidateSet? (this is my use-case atm)

**Comment by lzybkr** [comment]
The point of `Position` is so you can omit the parameter name, so in a way, it's like the opposite of what you want for Intellisense.

That said, dynamic validate set seems to be the primary use case for dynamic paramaters these days - so I opened an issue: https://github.com/PowerShell/PowerShell/issues/3744

**Comment by lfshr** [comment]
Closed as #3744 is a better solution.
